src/server/dflycmd.cc

```diff
@@ -305,13 +333,39 @@ OpStatus DflyCmd::StartFullSyncInThread(FlowInfo* flow, EngineShard* shard) {
   flow->saver.reset(new RdbSaver(flow->conn->socket(), save_mode, false));
 
   if (shard != nullptr) {
-    flow->saver->StartSnapshotInShard(false, shard);
+    auto ec = sf_->journal()->OpenInThread(false, string_view());
+    CHECK(!ec);
+    flow->saver->StartSnapshotInShard(true, shard);
   }
 
   flow->fb = ::boost::fibers::fiber(&DflyCmd::FullSyncFb, this, flow);
   return OpStatus::OK;
 }
 
+OpStatus DflyCmd::StartStableSyncInThread(FlowInfo* flow, EngineShard* shard) {
+  if (shard != nullptr) {
+    flow->saver->StopSnapshotInShard(shard);
+  }
+
+  // Wait for full sync to finish.
+  if (flow->fb.joinable()) {
+    flow->fb.join();
+  }
+
+  if (shard != nullptr) {
+    flow->saver.reset();
```

Comment:
I wonder if flow may be in an error state by this time. should we stop switching if we are already in error state?

Reply:
I don't propagate errors yet. I initially planned adding this to the cancellation PR. Should it be here instead?